File: CsharpProjects/myLeetCode/Dynamic_Programming/Basic/62. Unique Paths.cs
// 62. Unique Paths
// https://leetcode.com/problems/unique-paths/

#include <iostream>
#include <vector>

int uniquePaths(int m, int n) {
    std::vector<std::vector<int>> dp(m, std::vector<int>(n, 1));

    for (int i = 1; i < m; i++) {
        for (int j = 1; j < n; j++) {
            dp[i][j] = dp[i-1][j] + dp[i][j-1];
        }
    }

    return dp[m-1][n-1];
}

int main() {
    std::vector<std::pair<int, int>> testCases = {{3, 7}, {3, 2}};
    for (auto& testCase: testCases) {
        std::cout << "Number of unique paths: " << uniquePaths(testCase.first, testCase.second) << std::endl;
    }
    
    return 0;
}

File: C++/代码随想录/Dynamic_Programming/Basic/62. Unique Paths.py
def uniquePaths(m, n):
    dp = [[1] * n for _ in range(m)]

    for i in range(1, m):
        for j in range(1, n):
            dp[i][j] = dp[i - 1][j] + dp[i][j - 1]

    return dp[m - 1][n - 1]

# 测试用例
print(uniquePaths(3, 7))  # 应该输出 28
print(uniquePaths(3, 2))  # 应该输出 3
